### ISSUE TITLE
nn.LSTM documentation: output description

### ISSUE THREAD

**Issue by el1jah1** [pytorch/pytorch#139582]
### 📚 The doc issue

Hi. Could you please check if there are any inaccuracies in the documentation of [nn.LSTM](https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html#torch.nn.LSTM) related to the main description and output:

### **main description**:
> If `proj_size > 0` is specified, LSTM with projections will be used. This changes the LSTM cell in the following way. First, the dimension of $h_{t}$ will be changed from `hidden_size` to `proj_size` (dimensions of $W_{hi}$ will be changed accordingly).
Second, the output hidden state of each layer will be multiplied by a learnable projection matrix: $h_{t}=W_{hr}h_{t}$.

From the description it may seem that the projection applies only to output hidden state (after the last step), but in fact, judging by the size of **output** , it is applied after each step.

Also it is probably necessary to indicate that the dimensions also change for matrices **W_hf|W_hg|W_ho|W_ii|W_if|W_ig|W_io** (not only for $W_{hi}$) as can be seen from the description **weight_ih_l[k]** и **weight_hh_l[k]**.

the description should probably be:
> If `proj_size > 0` is specified, LSTM with projections will be used. This changes the LSTM cell in the following way.
The dimension of hidden state of each step of each layer will be changed from `hidden_size` to `proj_size` by multiplying by a learnable projection matrix: $h_{t}=W_{hr}h_{t}$ (dimensions of other matrices will be changed accordingly).

### **h_0**, **c_0**:
> tensor of shape ... for unbatched input or ... containing the final hidden state for each element in the sequence.

In fact, it does not contain a hidden state for each element of the sequence (only for the last element, that is, the last step).

> When bidirectional=True, h_n will contain a concatenation of the final forward and reverse hidden states, respectively.

Perhaps it would be better to clarify that the concatenation is along the 1st axis so that it is not analogous to the concatenation for `output` where the output directions are concatenated for each step:
> When bidirectional=True, h_n will contain a concatenation (along 1st axis) of the final forward and reverse hidden states, respectively.

or remove word 'concatenation' since it doesn't actually happen (the hidden vectors of the forward and backward passes are simply enumerated along the 1st axis)

**Comment by adelie-sokpha** [comment]
Hello I'm new to contributing in PyTorch, could I take this issue ?

**Comment by el1jah1** [comment]
Hi. Is this a question for me? If yes, then sure.

**Comment by adelie-sokpha** [comment]
Thanks !

1) > From the description it may seem that the projection applies only to output hidden state (after the last step), but in fact, judging by the size of **output**, it is applied after each step.

Based on what I’ve seen, the projection is applied after every step, so I think this part can be updated.

2) > Also it is probably necessary to indicate that the dimensions also change for matrices **W_hf|W_hg|W_ho|W_ii|W_if|W_ig|W_io** (not only for  𝑊 ℎ𝑖) as can be seen from the description **weight_ih_l[k]** и **weight_hh_l[k]**.

From what I’ve noticed, there is indeed a change in dimensions for the other matrices, as you mentioned, since w_hh changes dimensions due to projection (and includes these matrices: W_hi, W_hf, W_hg, W_ho). However, I think the overall description is correct, though it might help to add more detail about these matrices for clarity.

As for the projection part, I think your suggestion works well.

3) 
> ### **h_0**, **c_0**:
> > tensor of shape ... for unbatched input or ... containing the final hidden state for each element in the sequence.
> 
> In fact, it does not contain a hidden state for each element of the sequence (only for the last element, that is, the last step).

I think you meant \( h_n \) and \( c_n \) instead of \( h_0 \) and \( c_0 \) since you’re referring to the output description. It’s true that the last step is retained, but I’m not sure “element” actually means step in the given description.

4) 
> Perhaps it would be better to clarify that the concatenation is along the 1st axis so that it is not analogous to the concatenation for `output` where the output directions are concatenated for each step.
> Or remove word 'concatenation' since it doesn't actually happen (the hidden vectors of the forward and backward passes are simply enumerated along the 1st axis) 

I understand the confusion between “concatenation” for the output and “concatenation” in \( h_n \), as they’re not the same type of concatenation. However, from what I’ve seen, the term “concatenation” is accurate since \( h_n \) is created using a `torch.stack`, which in this case concatenates two tensors. 

I think the current description is correct, but adding a bit more detail—like you suggested—could help, such as specifying that it’s along the first axis.

**Comment by guest-oo** [comment]
Is the output h_n valid data masked by pack?

And the data structure of h_n is independent of whether batch_first=True？
Bidirectional GRU if i choose the final hiddeni state output in both directions is h-f-t, combined with h-b-1?
